### PR TITLE
#450 - Bulk run workflows

### PR DESCRIPTION
**Pull request by jozefbakus**
Implements: #450 

- Added confirmation dialog
- Limit number of workflows
- Logging for bulk workflows run
- New property: application.maximumNumberOfWorkflowsInBulkRun

### CODE REVIEW

File: src/main/scala/za/co/absa/hyperdrive/trigger/api/rest/services/WorkflowService.scala [hunk 1]
```diff
@@ -145,13 +164,22 @@ class WorkflowServiceImpl(override val workflowRepository: WorkflowRepository,
...
-      true
+    workflowIds.distinct.length match {
+      case numberOfWorkflows if numberOfWorkflows == 1 =>
+        throw new ApiException(GenericError(s"More than 1 workflow has to be triggered!"))
+      case numberOfWorkflows if numberOfWorkflows > ApplicationConfig.maximumNumberOfWorkflowsInBulkRun =>
+        throw new ApiException(GenericError(
+          s"Maximum number of workflows = ${ApplicationConfig.maximumNumberOfWorkflowsInBulkRun}, for bulk run has been exceeded!"
+        ))
+      case numberOfWorkflows =>
+        serviceLogger.debug(s"Bulk run workflows. ${numberOfWorkflows} workflows will be executed.")
```
Comment (kevinwallimann) [inline src/main/scala/za/co/absa/hyperdrive/trigger/api/rest/services/WorkflowService.scala]: Please also log the username and all workflow ids

File: src/main/scala/za/co/absa/hyperdrive/trigger/scheduler/utilities/Configs.scala [hunk 2]
```diff
@@ -125,3 +125,8 @@ object HealthConfig {
   lazy val yarnConnectionTestEndpoint: String =
     Configs.conf.getString("health.yarnConnection.testEndpoint")
 }
+
+object ApplicationConfig {
+  lazy val maximumNumberOfWorkflowsInBulkRun: Int =
```
Comment (kevinwallimann) [inline src/main/scala/za/co/absa/hyperdrive/trigger/scheduler/utilities/Configs.scala]: I think it makes no sense to use `lazy` here. I don't know why I used it in the `HealthConfig`

File: ui/src/app/components/workflows/workflows-home/workflows-home.component.html [hunk 3]
```diff
@@ -90,7 +90,7 @@ <h3 class="modal-title">Upload workflow</h3>
           <clr-icon shape="power" class="has-badge"></clr-icon>
           Deactivate
         </div>
-        <div aria-label="Run" clrDropdownItem (click)="runSelectedWorkflows(selected)">
+        <div aria-label="Run" clrDropdownItem [clrDisabled]="isRunSelectedWorkflowsDisabled(selected)" (click)="runSelectedWorkflows(selected)">
```
Comment (kevinwallimann) [inline ui/src/app/components/workflows/workflows-home/workflows-home.component.html:93]: It's disabled, but still clickable

File: src/main/scala/za/co/absa/hyperdrive/trigger/persistance/DagInstanceRepository.scala [hunk 4]
```diff
@@ -51,6 +53,17 @@ class DagInstanceRepositoryImpl extends DagInstanceRepository {
     }.transactionally
   ).map(_ => (): Unit)
 
+  override def insertJoinedDagInstances(dagInstancesJoined: Seq[DagInstanceJoined])(implicit executionContext: ExecutionContext): Future[Unit] = db.run(
+    DBIO.sequence {
+      dagInstancesJoined.map { dagInstanceJoined =>
+        for {
```
Comment (kevinwallimann) [inline src/main/scala/za/co/absa/hyperdrive/trigger/persistance/DagInstanceRepository.scala]: Please extract this duplicated code fragment

File: src/main/scala/za/co/absa/hyperdrive/trigger/api/rest/services/WorkflowService.scala [hunk 5]
```diff
@@ -143,15 +162,24 @@ class WorkflowServiceImpl(override val workflowRepository: WorkflowRepository,
...
-      dagInstanceJoined <- dagInstanceService.createDagInstance(joinedWorkflow.dagDefinitionJoined, userName)
-      _ <- dagInstanceRepository.insertJoinedDagInstance(dagInstanceJoined)
-    } yield {
-      true
+    workflowIds.distinct.length match {
+      case numberOfWorkflows if numberOfWorkflows == 1 =>
+        throw new ApiException(GenericError(s"More than 1 workflow has to be triggered!"))
+      case numberOfWorkflows if numberOfWorkflows > ApplicationConfig.maximumNumberOfWorkflowsInBulkRun =>
+        throw new ApiException(GenericError(
+          s"Maximum number of workflows = ${ApplicationConfig.maximumNumberOfWorkflowsInBulkRun}, for bulk run has been exceeded!"
```
Comment (kevinwallimann) [inline src/main/scala/za/co/absa/hyperdrive/trigger/api/rest/services/WorkflowService.scala]: The error message seems a bit cumbersome. How about e.g. "Cannot trigger more than x workflows"?

File: src/main/scala/za/co/absa/hyperdrive/trigger/api/rest/services/WorkflowService.scala [hunk 6]
```diff
@@ -143,15 +162,24 @@ class WorkflowServiceImpl(override val workflowRepository: WorkflowRepository,
...
+        ))
+      case numberOfWorkflows =>
+        serviceLogger.debug(s"Bulk run workflows. ${numberOfWorkflows} workflows will be executed.")
+        for {
+          joinedWorkflows <- workflowRepository.getWorkflows(workflowIds.distinct)
+          dagInstanceJoined <- Future.sequence(joinedWorkflows.map(joinedWorkflow => dagInstanceService.createDagInstance(joinedWorkflow.dagDefinitionJoined, userName)))
```
Comment (kevinwallimann) [inline src/main/scala/za/co/absa/hyperdrive/trigger/api/rest/services/WorkflowService.scala]: This line is a bit long

File: ui/src/app/stores/workflows/workflows.effects.spec.ts [hunk 7]
```diff
@@ -906,6 +907,95 @@ describe('WorkflowsEffects', () => {
...
+        },
+      });
+      expect(underTest.workflowsRun).toBeObservable(expected);
+      expect(toastrServiceSpy).toHaveBeenCalledTimes(1);
+      expect(toastrServiceSpy).toHaveBeenCalledWith(texts.RUN_WORKFLOWS_FAILURE_NOTIFICATION);
+    });
+
+    it('should display failure when service throws an exception while running workflows', () => {
+      const toastrServiceSpy = spyOn(toastrService, 'error');
+      const workflowId = 42;
```
Comment (kevinwallimann) [inline ui/src/app/stores/workflows/workflows.effects.spec.ts]: unused

File: ui/src/app/components/workflows/workflows-home/workflows-home.component.spec.ts [hunk 8]
```diff
@@ -380,6 +386,54 @@ describe('WorkflowsHomeComponent', () => {
...
+    waitForAsync(() => {
+      const subject = new Subject<boolean>();
+      spyOn(confirmationDialogService, 'confirm').and.returnValue(subject.asObservable());
+      subject.next(true);
+
+      const workflows = [
+        WorkflowModelFactory.create('workflowName1', true, 'projectName1', new Date(Date.now()), new Date(Date.now()), 0),
+        WorkflowModelFactory.create('workflowName2', true, 'projectName2', new Date(Date.now()), new Date(Date.now()), 1),
+      ];
+      const workflowIds = workflows.map((workflow) => workflow.id);
```
Comment (kevinwallimann) [inline ui/src/app/components/workflows/workflows-home/workflows-home.component.spec.ts]: unused

Comment (jozefbakus) [reply ui/src/app/components/workflows/workflows-home/workflows-home.component.spec.ts]: Ok

File: ui/src/app/stores/workflows/workflows.effects.spec.ts (hunk 7 above)
Comment (jozefbakus) [reply ui/src/app/stores/workflows/workflows.effects.spec.ts]: Ok

File: src/main/scala/za/co/absa/hyperdrive/trigger/api/rest/services/WorkflowService.scala (hunk 6 above)
Comment (jozefbakus) [reply src/main/scala/za/co/absa/hyperdrive/trigger/api/rest/services/WorkflowService.scala]: Ok

File: src/main/scala/za/co/absa/hyperdrive/trigger/api/rest/services/WorkflowService.scala (hunk 5 above)
Comment (jozefbakus) [reply src/main/scala/za/co/absa/hyperdrive/trigger/api/rest/services/WorkflowService.scala]: Ok

File: src/main/scala/za/co/absa/hyperdrive/trigger/persistance/DagInstanceRepository.scala (hunk 4 above)
Comment (jozefbakus) [reply src/main/scala/za/co/absa/hyperdrive/trigger/persistance/DagInstanceRepository.scala]: Ok

File: ui/src/app/components/workflows/workflows-home/workflows-home.component.html (hunk 3 above)
Comment (jozefbakus) [reply ui/src/app/components/workflows/workflows-home/workflows-home.component.html:93]: Ok

File: src/main/scala/za/co/absa/hyperdrive/trigger/scheduler/utilities/Configs.scala (hunk 2 above)
Comment (jozefbakus) [reply src/main/scala/za/co/absa/hyperdrive/trigger/scheduler/utilities/Configs.scala]: ok

File: src/main/scala/za/co/absa/hyperdrive/trigger/api/rest/services/WorkflowService.scala (hunk 1 above)
Comment (jozefbakus) [reply src/main/scala/za/co/absa/hyperdrive/trigger/api/rest/services/WorkflowService.scala]: ok

**Review by kevinwallimann — APPROVED** [review]

**Pull request merged**